### PR TITLE
Issue 1538: Expose MetricsServer in http server

### PR DESCRIPTION
**Pull request by jiazhai**
Descriptions of the changes in this PR:

Expose MetricsServer in  http server under `/metrics`

> ---
> In order to uphold a high standard for quality for code contributions, Apache BookKeeper runs various precommit
> checks for pull requests. A pull request can only be merged when it passes precommit checks. However running all
> the precommit checks can take a long time, some trivial changes don't need to run all the precommit checks. You
> can check following list to skip the tests that don't need to run for your pull request. Leave them unchecked if
> you are not sure, committers will help you:
>
> - [X] [skip bookkeeper-server bookie tests]: skip testing `org.apache.bookkeeper.bookie` in bookkeeper-server module.
> - [X] [skip bookkeeper-server client tests]: skip testing `org.apache.bookkeeper.client` in bookkeeper-server module.
> - [X] [skip bookkeeper-server replication tests]: skip testing `org.apache.bookkeeper.replication` in bookkeeper-server module.
> - [X] [skip bookkeeper-server tls tests]: skip testing `org.apache.bookkeeper.tls` in bookkeeper-server module.
> - [X] [skip bookkeeper-server remaining tests]: skip testing all other tests in bookkeeper-server module.
> - [ ] [skip integration tests]: skip docker based integration tests. if you make java code changes, you shouldn't skip integration tests.
> - [ ] [skip build java8]: skip build on java8. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> - [ ] [skip build java9]: skip build on java9. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> ---

### CODE REVIEW

**Comment by eolivelli** [comment]
There is a lot of code reformat, which is the real cause of the problem ?

File: bookkeeper-http/http-server/src/main/java/org/apache/bookkeeper/http/HttpRouter.java [hunk 1]
```diff
@@ -29,74 +29,76 @@
...
-  private final Map<String, Handler> endpointHandlers = new HashMap<>();
+    private final Map<String, Handler> endpointHandlers = new HashMap<>();
 
-  public HttpRouter(AbstractHttpHandlerFactory<Handler> handlerFactory) {
-    this.endpointHandlers.put(HEARTBEAT, handlerFactory.newHandler(HttpServer.ApiType.HEARTBEAT));
-    this.endpointHandlers.put(SERVER_CONFIG, handlerFactory.newHandler(HttpServer.ApiType.SERVER_CONFIG));
+    public HttpRouter(AbstractHttpHandlerFactory<Handler> handlerFactory) {
+        this.endpointHandlers.put(HEARTBEAT, handlerFactory.newHandler(HttpServer.ApiType.HEARTBEAT));
+        this.endpointHandlers.put(SERVER_CONFIG, handlerFactory.newHandler(HttpServer.ApiType.SERVER_CONFIG));
+        this.endpointHandlers.put(METRICS, handlerFactory.newHandler(HttpServer.ApiType.METRICS));
```
Comment (eolivelli) [inline bookkeeper-http/http-server/src/main/java/org/apache/bookkeeper/http/HttpRouter.java]: I have found, this is the real fix.
Can we add tests ?

**Comment by jiazhai** [comment]
@eolivelli, right, there were 2 commits, one for the real change. one for format.

**Comment by eolivelli** [comment]
@jiazhai sorry, I didn't want you to waste time in reverts.

GitHub has an awesome feature to "ignore white spaces" and so the diff looks clean anyway, this is why I found the real fix.
I wonder why checkstyle is okay with that file.

I think your re-format was good.
Feel free to apply it.

For me the important implementation to do is to add a test case

**Comment by eolivelli** [comment]
Second question:
is is not possible to use '/metrics' instead of /api/v1/metrics 

this looks difficult for the user for Prometheus integration

File: bookkeeper-http/http-server/src/main/java/org/apache/bookkeeper/http/HttpRouter.java [hunk 2]
```diff
@@ -32,6 +32,7 @@
   // Define endpoints here.
   public static final String HEARTBEAT                    = "/heartbeat";
   public static final String SERVER_CONFIG                = "/api/v1/config/server_config";
+  public static final String METRICS                      = "/api/v1/metrics";
```
Comment (sijie) [inline bookkeeper-http/http-server/src/main/java/org/apache/bookkeeper/http/HttpRouter.java]: it should be "/metrics".

File: site/docs/latest/admin/http.md [hunk 3]
```diff
@@ -60,6 +60,19 @@ Currently all the HTTP endpoints could be divided into these 4 components:
         |403 | Permission denied |
         |404 | Not found |
 
+## Config
+
+### Endpoint: /api/v1/metrics
```
Comment (sijie) [inline site/docs/latest/admin/http.md]: "/metrics"

**Comment by jiazhai** [comment]
retest this please

**Comment by jiazhai** [comment]
retest this please

**Comment by eolivelli** [comment]
rerun bookie tests

**Comment by sijie** [comment]
@jiazhai can you rebase the PR to latest master? so it doesn't run those flaky EntryLogTest tests.

**Comment by eolivelli** [comment]
retest this please

**Comment by sijie** [comment]
rerun bookkeeper-server bookie tests

**Comment by jiazhai** [comment]
@sijie merged with latest master. will wait for the result.

**Comment by sijie** [comment]
cherry-picked the change to 4.7.3

**Pull request closed without merging**